domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgress.kt

```diff
@@ -47,6 +47,22 @@ internal class QuestionAssessmentProgress {
     questionSessionMetrics[currentQuestionIndex].didViewSolution = true
   }
 
+  /**
+   * Called every time the user submits an answer to track how many answers the user submitted per question, along with
+   * any misconceptions.
+   */
+  internal fun answerSubmitted(taggedSkillMisconceptionId: String?) {
+    val currentQuestionIndex = getCurrentQuestionIndex()
+    questionSessionMetrics[currentQuestionIndex].numberOfAnswersSubmitted++
+    if (taggedSkillMisconceptionId != null) {
+      // Misconceptions are formatted as <skill ID>-<misconception ID> from the AnswerClassificationController,
```

Comment:
Is this something we did, or is it formatted that way per the exploration structure? If the former, we ought to use a data class to extract the values rather than string formatting (e.g. to avoid the [stringly typed antipattern](https://wiki.c2.com/?StringlyTyped)).

Reply:
I believe this is formatted per the exploration structure - I checked the web repo and [this](https://github.com/oppia/oppia/blob/3fcc934d20cd756117a56b361315739fc687606d/core/templates/pages/exploration-editor-page/editor-tab/templates/modal-templates/add-answer-group-modal.controller.ts#L60) is how they're creating AnswerGroups. If you look at how they manage the mastery/score calculation logic, [this](https://github.com/oppia/oppia/blob/98c62ff192d75a0ffdaea24ec6559cc4be5a932b/core/templates/components/question-directives/question-player/question-player.directive.ts#L408) is exactly how they do it as well.

Reply:
I think this is an area for potential improvement, then. Could we potentially replace the 'string' in our proto structure with a structured alternative the represents the constituent pieces? For example, we could have a structured alternative if we used something like:

```proto
message Misconception {
  string skill_id = 1;
  string misconception_id = 2;
}
```

It will require us to do this parsing when we retrieve the JSON object and convert it to proto (e.g. in StateRetriever), but that's already a location that's responsible for parsing so it's not unexpected, and it facilitates leveraging a structured approach elsewhere in the codebase which can help minimize error. WDYT?

FWIW, I suggest following up with a code change to implement the above rather than pulling it in here, but I think it would be a very nice improvement to avoid the antipattern being introduced.

Reply:
Yeah I think that's definitely a better approach, I'll go ahead and make an issue for that. Thanks!